src/viam/sdk/common/linear_algebra.hpp

```diff
@@ -0,0 +1,60 @@
+#pragma once
+#include <boost/qvm/vec.hpp>
+#include <boost/qvm/vec_traits.hpp>
+
+#include <viam/api/common/v1/common.pb.h>
+namespace viam {
+namespace sdk {
+
+class Vector3 {
+   public:
+    Vector3(double x, double y, double z) : data_{x, y, z} {};
+    Vector3() : Vector3(0.0, 0.0, 0.0){};
+
+    double x() const;
+    double y() const;
+    double z() const;
+    void set_x(double x);
+    void set_y(double y);
+    void set_z(double z);
+    std::array<double, 3> data() const;
+    std::array<double, 3>& mutable_data();
```

Comment:
From a language perspective, that is super cool. 

I talked with @stuqdog about this and right now I think the desired style is towards explicit mutability but I will keep this in mind going forward.

Reply:
Hmm. OK. I'd say that is somewhat non-idiomatic for C++. Overloading accessors on `const` is pretty much the way it is done. It has the advantage that I don't need to know whether the object I'm operating on is `const` or not to call `data`. I get the right behavior and the right return type either way.

For instance, `std::vector` overloads `begin` such that when called on a non-const `std::vector` it returns a `std::vector::iterator` and when called on a `const` std::vector it returns a `std::vector::const_iterator`: https://en.cppreference.com/w/cpp/container/vector/begin. Don't be misled by the `cbegin` stuff that you will see there: that is there as a convenience shortcut to explicitly get a `const_iterator` when you may not have a const object, saving you a `const_cast` to `const`.

Basically, you want `const` to "flow": const objects return const references which in turn allows calling const methods. Making the name different means there is a point of divergence where you need to know the constness to know which method to call.

But I'll leave it to Ethan to make the final call.

Reply:
@stuqdog - I would like to hear your thoughts on my comments above.

Reply:
@acmorrow your explanation makes sense and I think I'm leaning toward being convinced that we should overload on `const`ness. My concern is that this will lead to cases where people get mutable objects when they don't need/want mutability. Even if one has a non-`const` `T`, I could imagine they'd still want `const` behavior from certain calls. They could cast, but this requires extra code and greater awareness of the `const`ness of values. 

Given your statement that overloading on `const` is the more idiomatic approach, perhaps I am just speaking from a place if ignorance or trying to force C++ into something it isn't by using non-idiomatic paradigms from different languages. I'm certainly willing to be convinced here!

Reply:
Moving forward with `const`-based overload of `data()`

Reply:
Just spoke with @acmorrow offline, I have been convinced that overloading on `const` is the right approach here.